### PR TITLE
Convert project to TypeScript

### PR DESCRIPTION
**Pull request by ComradeVanti**
This is an initial attempt to transition the project to TypeScript.

The tests pass, but obviously there would need to be an additional compile step somewhere when building the project for production.

This conversion has exposed **a lot** of potential `undefined` and other errors. I have highlighted these with `TODO` comments.

Hope to get some feedback on this.

### CODE REVIEW

**Comment by favoyang** [comment]
Wow, this is really cool. I appreciate TypeScript for its debugging ease. There's a lot of work ahead, and I admire your passion and efforts. Thanks for that. I might have more feedback to share over the weekend.

Currently, our CI is encountering issues due to an incompatible Node version. The project requires Node v14, as specified in the `.nvmrc` file, while it seems you are using Node v18. I hope we can support v14 to support a smoother upgrade for our existing users. However, if it's necessary to work with the latest LTS, I'm open to that too.

**Comment by ComradeVanti** [comment]
Oops, didn't notice the node change. I might take a look to see if I can rework it for v14 if you thinks that's better. Let me know what you think.

**Comment by ComradeVanti** [comment]
Just did a rebase to include 23f43fdab02a9b9cd84ee2b6a10cf1d358fc4e7d. Hope that is what you meant

**Comment by favoyang** [comment]
Yes, the CI is passed now.

**Comment by ComradeVanti** [comment]
Cool. Do you need any more changes?

**Comment by favoyang** [comment]
No, thanks. Perhaps I will give more feedback on the weekend.

**Comment by favoyang** [comment]
Sure you can do more refactors.

**Comment by ComradeVanti** [comment]
Cool, I'll try to keep it at refactors only, or at maximum add more tests. But no new features for now. Ulitmately I'd like to add https://github.com/openupm/openupm-cli/issues/49 ;)

**Comment by favoyang** [comment]
> @favoyang could you help me understand a bit of your code? Specifically it's this bit right here inside of the `getLatestVersion` function.
> 
> ```js
> return Object.keys(pkgInfo.versions).find(
>   (key) => pkgInfo.versions[key] == "latest"
> );
> ```
> 
> This function is supposed to return a version-name, aka a `string`. I don't see how this code would work, when looking at the structure of an example package, such as this one: image It seems like `versions` is an object where every key is a version name. The values are the version-objects. So when you say `pkgInfo.versions[key] == "latest"` aren't you attempting to equate a whole version-object with a `string`? If we want to find the latest version, shouldn't we instead just do `Object.keys(pkgInfo.versions).pop();` in order to get the last version in the object?

Regarding the `test-core.js`, it seems I was attempting to handle a legacy data format that might be present on the Verdaccio server or the Unity registry. However, for now, we can simplify things and assume that the only way to retrieve the latest version is by reading from the `"dist-tags"` field. You can safely remove the old logic and the associated test cases.

```
  describe("getLatestVersion", function() {
    it("from dist-tags", async function() {
      getLatestVersion({ "dist-tags": { latest: "1.0.0" } }).should.equal(
        "1.0.0"
      );
    });
    it("from versions", async function() {
      getLatestVersion({ versions: { "1.0.0": "latest" } }).should.equal(
        "1.0.0"
      );
    });
    it("not found", async function() {
      (
        getLatestVersion({ versions: { "1.0.0": "patch" } }) === undefined
      ).should.be.ok();
      (getLatestVersion({}) === undefined).should.be.ok();
    });
  });
```

**Comment by ComradeVanti** [comment]
@favoyang I'm done fixing all TS errors while also ensuring all tests pass. It could still be that I broke something somewhere. Would love to refactor further but kept it at a minimum for now. Let me know if you have any questions.

**Comment by ComradeVanti** [comment]
For the record 8027f16c0342c406851a88ba2ea1fad6b1c0af7f is the commit at which the project is fully converted to TS but no further major refactors have taken place.

**Comment by ComradeVanti** [comment]
Another update. I have now fully converted the tests to TS as well. The commit for that is 33a26741aa7d3a7e6d65fb8e657a281d2319eec0

**Comment by ComradeVanti** [comment]
@favoyang another important update. I changed the compile/run structure of the package to work better with typescript. I describe in more detail in f88bc8b1edef5e6b2add9ac140d190eb65aa568a. Also I just wanted to know if you are interested in merging this at some point. It has been a fun exercise for me, but if you would rather not switch over, I'll not invest more time and leave it here. Just let me know :)

**Comment by favoyang** [comment]
> Also I just wanted to know if you are interested in merging this at some point. It has been a fun exercise for me, but if you would rather not switch over, I'll not invest more time and leave it here. Just let me know :)

Yes, I like the idea, and I will spend some time reviewing it soon. The last weekend was quite exhausting due to another major refactor of the main website, which you can find at https://github.com/openupm/openupm/issues/4042.

By the way, it might be a good idea to freeze this PR unless some bug fixes are made.

Thanks for your patience.

**Comment by ComradeVanti** [comment]
No worries, thanks for responding. Alright, I'll leave it for now. I have started extracting some utility functions, but not much, so that looks a little bit out of place now. Let me know if you have any questions.

File: package.json [hunk 1]
```diff
@@ -35,8 +35,23 @@
...
+    "@types/mocha": "^10.0.3",
+    "@types/node": "^14.18.63",
+    "@types/npmlog": "^4.1.4",
+    "@types/pkginfo": "^0.4.1",
+    "@types/promptly": "^3.0.3",
+    "@types/request": "^2.48.11",
+    "@types/should": "^13.0.0",
+    "@types/test-console": "^2.0.1",
+    "@types/update-notifier": "^6.0.5",
+    "@typescript-eslint/eslint-plugin": "^6.8.0",
```
Comment (favoyang) [inline package.json:51]: I received a few warnings for `EBADENGINE` when running `npm install`. All of them are related to `@typescript-eslint` or its dependencies, such as `ts-api-utils`. Initially, I thought it was related to ESLint, which is a dev dependency and should not affect users who install the package. However, even running `npm install --omit=dev` resulted in the same error.

```plaintext
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@typescript-eslint/eslint-plugin@6.8.0',
npm WARN EBADENGINE   required: { node: '^16.0.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v14.19.2', npm: '9.6.7' }
npm WARN EBADENGINE }
...

npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'ts-api-utils@1.0.3',
npm WARN EBADENGINE   required: { node: '>=16.13.0' },
npm WARN EBADENGINE   current: { node: 'v14.19.2', npm: '9.6.7' }
npm WARN EBADENGINE }
```

It appears that the development environment requires Node.js v16. This implies that we should develop and test everything in Node.js v16. To address this, we might consider requiring an upgrade to Node.js v16, which can be done by updating the .nvmrc file and the `engines` field in the package.json.

File: lib/index.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env -S node -r "ts-node/register"
```
Comment (favoyang) [inline lib/index.ts:1]: Just curious ts-node vs. compiling, which approach is better? I appreciate that ts-node simplifies the development workflow, but I'm uncertain about the trade-offs.

**Review by favoyang — CHANGES_REQUESTED** [review]
LGTM. Some minor discussions were added.

**Comment by favoyang** [comment]
Also, please note that there's a conflict due to the merging of #50.

**Comment by ComradeVanti** [comment]
> Also, please note that there's a conflict due to the merging of #50.

I hope to have fixed that now.

**Comment by favoyang** [comment]
> > Also, please note that there's a conflict due to the merging of #50.
> 
> I hope to have fixed that now.

Thanks for the rebase. I think my comments have collapsed here: https://github.com/openupm/openupm-cli/pull/52#pullrequestreview-1702992783, please checkout.

**Comment by ComradeVanti** [comment]
> Thanks for the rebase. I think my comments have collapsed here: [#52 (review)](https://github.com/openupm/openupm-cli/pull/52#pullrequestreview-1702992783), please checkout.

Yeah sorry, I force pushed and now all the commits are here twice. But I already responded to your feedback.

**Comment by favoyang** [comment]
> But I already responded to your feedback.

@ComradeVanti, sorry I didn't see your response for my previous review https://github.com/openupm/openupm-cli/pull/52#pullrequestreview-1702992783?

image

**Comment by ComradeVanti** [comment]
@favoyang That's weird. Here is how it looks on my side:
image
Maybe I'm doing something wrong?

**Comment by favoyang** [comment]
It says pending. Perhaps there is a button for submitting.

Refs https://github.com/orgs/community/discussions/10369

> After much looking, I found the place to "Submit" my review on the "files changed" tab in the upper right hand corner.

**Comment by ComradeVanti** [comment]
Oh god I feel stupid. Thanks, I hope I did it now.

File: package.json (hunk 1 above)
Comment (favoyang) [reply package.json:51]: Please also check if the README mentions the need for Node v16.

Comment (ComradeVanti) [reply package.json:51]: Alright will do

Comment (ComradeVanti) [reply package.json:51]: See d53b70b9b7791cda87b4b8c9260e23176a59fbce

File: lib/index.ts (hunk 2 above)
Comment (ComradeVanti) [reply lib/index.ts:1]: Ok yes, that makes more sense. I'll do that.

Comment (ComradeVanti) [reply lib/index.ts]: See 3c32160e5a7738f82b4a550bf4d391b70402d094 and 5903497600fd3f0c07ff1e7a1851b3aeee22502f. Specifically for the CI in 5903497600fd3f0c07ff1e7a1851b3aeee22502f I'm not sure if that is what you meant, please let me know. Thanks.

File: lib/client.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      actual: x,
+    });
+}
+
+/**
+ * Normalizes a RegClient function. Specifically it merges it's multiple
+ * callback arguments into a single NormalizedError object. This function
+ * also takes care of binding and promisifying.
+ */
+function normalizeClientFunction<TParam, TData>(
```
[the remark this answers is not in the document]
Comment (favoyang) [reply lib/client.ts]: Yes, promisify ignore callback parameters other than (err, result).

Before the status and code has been assigned by [makeError](https://github.com/openupm/another-npm-registry-client/blob/ca9cddded522819a9dcb4cb0317f6d191c9980f2/lib/request.js#L328C10-L328C20), so I don't need to access the last response parameter.

Anyway, it sounds okay here., at least we get a typed NpmClientError object that is easier to handle later.

File: package.json (hunk 1 above)
Comment (favoyang) [reply package.json:52]: https://github.com/openupm/openupm-cli/commit/d53b70b9b7791cda87b4b8c9260e23176a59fbce LGTM

File: lib/index.ts (hunk 2 above)
Comment (favoyang) [reply lib/index.ts]: All GLTM.

**Comment by favoyang** [comment]
One more thing, please remove ts-node from the dependencies (or move to devDeps if still needed) and regenerate the package-lock.json.

**Comment by ComradeVanti** [comment]
@favoyang Done. 1fae8e1bb22263aff2e59ff0e39ff5ea4668d0e9 and a62fb6148a91b735e075e638ef5f3d41ade008e8

**Review by favoyang — APPROVED** [review]

**Comment by favoyang** [comment]
Sounds like we're good here. I will give a few more manual tests and merge them tomorrow.

**Comment by ComradeVanti** [comment]
Cool thanks. I'm a big fan of this project an plan to contribute more. Looking forward to some great collaboration.

**Comment by favoyang** [comment]
@ComradeVanti thanks again for all the passion. I have added you to the contributor list.

https://github.com/openupm/openupm-cli#contributors

**Pull request merged**